core/blockchain.go

```diff
@@ -857,6 +867,9 @@ func (self *BlockChain) InsertChain(chain types.Blocks) (int, error) {
 				continue
 			}
 
+			self.futureMu.Lock()
+			defer self.futureMu.Unlock()
```

Comment:
Ouch. Nothing inherently wrong locking within the loop though. It's the unlocking which is problematic.